keps/sig-storage/1845-prioritization-on-volume-capacity/README.md

```diff
@@ -0,0 +1,424 @@
+achieve better storage resource usage.
+
+See https://github.com/kubernetes/kubernetes/issues/83323.
+
+#### Different weights for different classes
+
+Given a cluster which has two kinds of local storage classes:
+
+- hdd
+- ssd
+
+Workloads may request two disks, one from `hdd` and one from `ssd`. As
+the ssd storage resource may be relatively scarce, the administrator can
+configure a relatively larger weight for storage class `ssd` and hope it's more
+balanced.
+
+This applies to any other storage with topology constraints.
+
+## Design Details
+
+We have a scheduler plugin called
+[VolumeBinding](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/framework/plugins/volumebinding)
+to filter the nodes based on volume topology constraints. After the filter
+phase, there is a reduced set of nodes that can fit a pod. We can implement
+`Score` interfaces to score nodes based on assumed PVs on each node.
+
+The score is calculated in two steps:
+
+1) First, we group pod's unbound PVCs and corresponding static provisioned PVs
+by storage class. For PVCs/PVs in each storage class, we calculate requested to
+capacity ratio using the following formula:
+
+```
+[UtilizationPerClass] = [MaxUtilization] * ([RequestPerClass] / [CapacityPerClass])
+```
+
+The utilization ranges from 0 to 100.
+
+We can tune this formula by applying utilization shape points, which is used in
+[`RequestedToCapacityRatio`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio.go).
+
+By default, we use this shape points to get linear scores between [0,
+100] utilization.
+
+```
+- utilization: 0
+  score: 0
+- utilization: 100
+  score: config.MaxCustomPriorityScore
+```
+
+The score ranges from 0 to `config.MaxCustomPriorityScore`. (Internal score
+will be scaled up to the score range used by the scheduler).
+
+Optionally, users can configure customized utilization shape points, e.g.
+
+```
+- utilization: 50
+  score: 0
+- utilization: 80
+  score: 3
+- utilization: 100
+  score: 5
+```
+
+The score of utilization below 50 will be 0, and the score of utilization of 90
+will be 4 (`3 + (5 - 3) * (90 - 80) / (100 - 80)`).
+
+2) Then, we add them together and divides by the number of storage classes.
+
+### Configuring the utilization shape points
+
+We can add a field `Shape` to specify the shape points which defines the score
+function shape.
+
+```
+// VolumeBindingArgs holds arguments used to configure the VolumeBinding plugin.
+type VolumeBindingArgs struct {
+		...
+    // Shape specifies the shape points defining score function shape
+    Shape []UtilizationShapePoint
+}
+```
+
+The definition of `UtilizationShapePoint` is [here](https://github.com/kubernetes/kubernetes/blob/73fa63a86d1d47d82b435ca85c2f350da06b08b9/pkg/scheduler/apis/config/types_pluginargs.go#L115-L121).
+
+### Configuring the weight of storage class
```

Comment:
I am probably missing something, but doesn't the PVC spec include storageClassName field that a PV should match on to be eligible for binding? or is it the case that a PVC can be bound to a PV with a different storageClassName?

Reply:
ok, this is now explained in your reply to the weight question.

Reply:
yes, they're specified in the `storageClassName` field. `weights` here are numeric multipliers in scoring, then some storage classes can have a higher weight in scoring. negligible storage class can have a lower priority or be completely ignored if the weight is set to `0`.